exp/addons/internal/controllers/clusterresourceset_scope.go

```diff
@@ -0,0 +1,185 @@
+	// hash returns a computed hash of the defined objects in the resource. It is consistent
+	// between runs.
+	hash() string
+}
+
+func reconcileScopeForResource(
+	crs *addonsv1.ClusterResourceSet,
+	resourceRef addonsv1.ResourceRef,
+	resourceSetBinding *addonsv1.ResourceSetBinding,
+	resource *unstructured.Unstructured,
+) (resourceReconcileScope, error) {
+	normalizedData, err := normalizeData(resource)
+	if err != nil {
+		return nil, err
+	}
+
+	objs, err := objsFromYamlData(normalizedData)
+	if err != nil {
+		return nil, err
+	}
+
+	return newResourceReconcileScope(crs, resourceRef, resourceSetBinding, normalizedData, objs), nil
+}
+
+func newResourceReconcileScope(
+	clusterResourceSet *addonsv1.ClusterResourceSet,
+	resourceRef addonsv1.ResourceRef,
+	resourceSetBinding *addonsv1.ResourceSetBinding,
+	normalizedData [][]byte,
+	objs []unstructured.Unstructured,
+) resourceReconcileScope {
+	base := baseResourceReconcileScope{
+		clusterResourceSet: clusterResourceSet,
+		resourceRef:        resourceRef,
+		resourceSetBinding: resourceSetBinding,
+		data:               normalizedData,
+		normalizedObjs:     objs,
+		computedHash:       computeHash(normalizedData),
+	}
+
+	switch addonsv1.ClusterResourceSetStrategy(clusterResourceSet.Spec.Strategy) {
+	case addonsv1.ClusterResourceSetStrategyApplyOnce:
+		return &reconcileApplyOnceScope{base}
+	case addonsv1.ClusterResourceSetStrategyReconcile:
+		return &reconcileStrategyScope{base}
+	default:
+		return nil
+	}
+}
+
+type baseResourceReconcileScope struct {
+	clusterResourceSet *addonsv1.ClusterResourceSet
+	resourceRef        addonsv1.ResourceRef
+	resourceSetBinding *addonsv1.ResourceSetBinding
+	normalizedObjs     []unstructured.Unstructured
+	data               [][]byte
+	computedHash       string
+}
+
+func (b baseResourceReconcileScope) objs() []unstructured.Unstructured {
+	return b.normalizedObjs
+}
+
+func (b baseResourceReconcileScope) hash() string {
+	return b.computedHash
+}
+
+type reconcileStrategyScope struct {
+	baseResourceReconcileScope
+}
+
+func (r *reconcileStrategyScope) needsApply() bool {
+	resourceBinding := r.resourceSetBinding.GetResource(r.resourceRef)
+
+	return resourceBinding == nil || !resourceBinding.Applied || resourceBinding.Hash != r.computedHash
+}
+
+func (r *reconcileStrategyScope) apply(ctx context.Context, c client.Client) error {
+	return apply(ctx, c, r.applyObj, r.objs())
+}
+
+func (r *reconcileStrategyScope) applyObj(ctx context.Context, c client.Client, obj *unstructured.Unstructured) error {
+	currentObj := &unstructured.Unstructured{}
+	currentObj.SetAPIVersion(obj.GetAPIVersion())
+	currentObj.SetKind(obj.GetKind())
+	err := c.Get(ctx, client.ObjectKeyFromObject(obj), currentObj)
+	if apierrors.IsNotFound(err) {
+		return createUnstructured(ctx, c, obj)
+	}
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"reading object %s %s",
+			obj.GroupVersionKind(),
+			klog.KObj(obj),
+		)
+	}
```

Comment:
```suggestion
	if err != nil {
		if apierrors.IsNotFound(err) {
			return createUnstructured(ctx, c, obj)
		}

		return errors.Wrapf(
			err,
			"reading object %s %s",
			obj.GroupVersionKind(),
			klog.KObj(obj),
		)
	}
```
I know it's the same behavior, but I think this way it would be slightly easier to understand

Reply:
Ah interesting
IMO the nested ifs make it a bit more difficult to read
What's the part that you think this improves? Maybe I can come up with something else

Reply:
I think I like that we're explicitly checking that err != nil and not just calling IsNotFound

Just feels more explicitly like: "if we have an error and if that error is not found"

```go
if err != nil && apierrors.IsNotFound(err) {
```


That would achieve the same.

I guess the nested if is mostly that I'm used to always reading `if err != nil {` after something returns an error and then in the following block the entire error handling is done

Reply:
Ah I see. Isn't "we have an error" redundant when saying "we have a not-found error"?
That said, what about making both checks part of an if-else block? This groups conceptually the two branches that handle errors:
```suggestion
	if apierrors.IsNotFound(err) {
		return createUnstructured(ctx, c, obj)
	} else if err != nil {
		return errors.Wrapf(
			err,
			"reading object %s %s",
			obj.GroupVersionKind(),
			klog.KObj(obj),
		)
	}
```

Reply:
I would prefer to avoid the else branch if possible. Given it's not necessary as we return in the first one
(I think we even have a linter for it :))

Btw, it's fine for me to keep the current version.